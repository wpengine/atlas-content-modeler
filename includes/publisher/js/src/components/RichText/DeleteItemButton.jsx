import React from "react";
import TrashIcon from "../../../../../components/icons/TrashIcon";
import { __ } from "@wordpress/i18n";

const DeleteItemButton = ({ index, setValues }) => {
	return (
		<button
			type="button"
			onClick={(event) => {
				event.preventDefault();
				// Removes the value at the given index.
				setValues((currentValues) => {
					const newValues = [...currentValues];
					newValues.splice(index, 1);
					return newValues;
				});
			}}
		>
			<a aria-label={__("Remove item.", "atlas-content-modeler")}>
				<TrashIcon size="small" />{" "}
			</a>
		</button>
	);
};

export default DeleteItemButton;
