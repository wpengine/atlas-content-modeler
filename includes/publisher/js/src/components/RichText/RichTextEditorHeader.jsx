import React from "react";

const RichTextEditorHeader = ({ modelSlug, field }) => {
	return (
		<>
			<label
				htmlFor={`atlas-content-modeler[${modelSlug}][${field.slug}]`}
			>
				{field.name}
			</label>
			{field?.description && (
				<p className="help mb-4">{field.description}</p>
			)}
		</>
	);
};

export default RichTextEditorHeader;
