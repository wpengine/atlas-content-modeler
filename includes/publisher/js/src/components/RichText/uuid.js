import { v4 as uuidv4 } from "uuid";

/**
 * Generates a unique ID for field initialization and keying.
 *
 * Exported here to allow mocking with Jest.
 *
 * @returns {string}.
 */
export function uuid() {
	return "field-" + uuidv4();
}
