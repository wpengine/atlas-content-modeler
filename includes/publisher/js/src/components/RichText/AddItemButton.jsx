/** @jsx jsx */
import { jsx, css } from "@emotion/react";
import AddIcon from "../../../../../components/icons/AddIcon";
import { colors } from "../../../../../shared-assets/js/emotion";
import { __ } from "@wordpress/i18n";
import { v4 as uuidv4 } from "uuid";

const AddItemButton = ({ setValues }) => {
	const addItem = () =>
		setValues((oldValues) => [...oldValues, { id: uuidv4(), value: "" }]);

	return (
		<button
			onClick={(event) => {
				event.preventDefault();
				addItem();
			}}
			type="button"
			css={css`
				align-items: center;
				background: transparent;
				border: none;
				color: ${colors.primary};
				cursor: pointer;
				display: flex;
				font-weight: bold;
				height: 68px;
				margin: 8px;
				svg {
					margin-right: 4px;
				}
				&:focus,
				&:hover {
					color: ${colors.primaryHover};
					svg {
						path {
							fill: ${colors.primaryHover};
						}
					}
				}
			`}
		>
			<AddIcon noCircle />{" "}
			<span>{__(`Add Item`, "atlas-content-modeler")}</span>
		</button>
	);
};

export default AddItemButton;
