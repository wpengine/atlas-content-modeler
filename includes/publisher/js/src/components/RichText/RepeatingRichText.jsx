/** @jsx jsx */
import { jsx, css } from "@emotion/react";
import React from "react";
import RichTextHeader from "./RichTextHeader";
import AddItemButton from "../shared/repeaters/AddItemButton";
import DeleteItemButton from "../shared/repeaters/DeleteItemButton";
import { colors } from "../../../../../shared-assets/js/emotion";

const RepeatingRichText = ({ modelSlug, field, values, setValues, uuid }) => {
	const addItem = () =>
		setValues((oldValues) => [...oldValues, { id: uuid(), value: "" }]);

	const deleteItem = (index) =>
		setValues((currentValues) => {
			const newValues = [...currentValues];
			newValues.splice(index, 1);
			return newValues;
		});

	return (
		<>
			<RichTextHeader modelSlug={modelSlug} field={field} />
			{values.map(({ id, value }, index) => {
				return (
					<div
						key={id}
						data-testid="rich-text-repeater-row"
						css={css`
							display: flex;
							align-items: center;
							border: solid 1px ${colors.border};
							padding: 12px;
							.classic-form .field & .mce-tinymce {
								width: 100% !important;
							}
							${index === 0 &&
							values.length > 1 &&
							`border-bottom: none;`}
							${values.length > 1 && `padding-right: 0`}
						`}
					>
						<span
							className="px-1 me-2"
							css={css`
								font-family: "Open Sans", sans-serif;
								font-weight: bold;
							`}
						>
							{index + 1}
						</span>
						<textarea
							name={`atlas-content-modeler[${modelSlug}][${field.slug}][${index}]`}
							id={id}
							defaultValue={value}
						/>
						{values.length > 1 && (
							<DeleteItemButton
								deleteItem={() => deleteItem(index)}
							/>
						)}
					</div>
				);
			})}
			<div
				css={css`
					border: solid 1px ${colors.border};
					border-top: none;
				`}
			>
				<AddItemButton addItem={addItem} />
			</div>
		</>
	);
};

export default RepeatingRichText;
