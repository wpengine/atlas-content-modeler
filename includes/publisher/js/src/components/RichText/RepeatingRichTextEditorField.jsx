/** @jsx jsx */
import { jsx, css } from "@emotion/react";
import RichTextEditorHeader from "./RichTextEditorHeader";
import AddItemButton from "./AddItemButton";
import DeleteItemButton from "./DeleteItemButton";
import { colors } from "../../../../../shared-assets/js/emotion";

const RepeatingRichTextEditorField = ({
	modelSlug,
	field,
	values,
	setValues,
}) => {
	return (
		<>
			<RichTextEditorHeader modelSlug={modelSlug} field={field} />
			{values.map(({ id, value }, index) => {
				return (
					<div
						key={id}
						css={css`
							display: flex;
							align-items: center;
							border: solid 1px ${colors.border};
							padding: 12px;
							.classic-form .field & .mce-tinymce {
								width: 100% !important;
							}
							${index === 0 &&
							values.length > 1 &&
							`border-bottom: none;`}
							${values.length > 1 && `padding-right: 0`}
						`}
					>
						<textarea
							name={`atlas-content-modeler[${modelSlug}][${field.slug}][${index}]`}
							id={id}
							defaultValue={value}
						/>
						{values.length > 1 && (
							<DeleteItemButton
								index={index}
								setValues={setValues}
							/>
						)}
					</div>
				);
			})}
			<div
				css={css`
					border: solid 1px ${colors.border};
					border-top: none;
				`}
			>
				<AddItemButton setValues={setValues} />
			</div>
		</>
	);
};

export default RepeatingRichTextEditorField;
