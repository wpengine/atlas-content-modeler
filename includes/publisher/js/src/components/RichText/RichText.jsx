import React, { useState } from "react";
import { uuid } from "./uuid";
import SoloRichText from "./SoloRichText";
import RepeatingRichText from "./RepeatingRichText";
import useWpEditor from "./useWPEditor";
import { __ } from "@wordpress/i18n";

export default function RichText({ field, modelSlug }) {
	const initialValues = field?.isRepeatableRichText
		? (field?.value || [""]).map((val) => {
				return { id: uuid(), value: val };
		  })
		: [{ id: uuid(), value: field?.value }];

	const [values, setValues] = useState(initialValues);

	let textareaIds = values.map(({ id }) => id);

	useWpEditor(textareaIds);

	return field?.isRepeatableRichText ? (
		<RepeatingRichText
			modelSlug={modelSlug}
			field={field}
			values={values}
			setValues={setValues}
			uuid={uuid}
		/>
	) : (
		<SoloRichText
			modelSlug={modelSlug}
			field={field}
			fieldId={values[0]["id"]}
		/>
	);
}
