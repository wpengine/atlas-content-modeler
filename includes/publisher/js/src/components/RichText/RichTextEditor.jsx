import React, { useState } from "react";
import { v4 as uuidv4 } from "uuid";
import SoloRichTextEditorField from "./SoloRichTextEditorField";
import RepeatingRichTextEditorField from "./RepeatingRichTextEditorField";
import useWpEditor from "./useWPEditor";
import { __ } from "@wordpress/i18n";

export default function RichTextEditor({ field, modelSlug }) {
	// Generates a unique ID for each rich text field for initialization and keying.
	const initialValues = field?.isRepeatableRichText
		? (field?.value || [""]).map((val) => {
				return { id: uuidv4(), value: val };
		  })
		: [{ id: uuidv4(), value: field?.value }];

	const [values, setValues] = useState(initialValues);

	useWpEditor(values);

	return field?.isRepeatableRichText ? (
		<RepeatingRichTextEditorField
			modelSlug={modelSlug}
			field={field}
			values={values}
			setValues={setValues}
		/>
	) : (
		<SoloRichTextEditorField
			modelSlug={modelSlug}
			field={field}
			fieldId={values[0]["id"]}
		/>
	);
}
