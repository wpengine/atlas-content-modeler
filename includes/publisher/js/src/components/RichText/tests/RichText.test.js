import React from "react";
import renderer from "react-test-renderer";
import RichText from "../RichText";
import * as uuid from "../uuid"; // Exposed so Jest can mock this function.

beforeEach(() => {
	jest.resetAllMocks();
});

describe("RichText", () => {
	it("renders a solo field if not repeatable", () => {
		const nonRepeatableProps = {
			modelSlug: "cats",
			field: {
				slug: "name",
				name: "Name",
				description: "The cat's real name.",
				value: "<p>Kahn The Destroyer</p>",
				isRepeatableRichText: false,
			},
		};
		// A dynamic ID would break snapshot testing, so we use a fixed one.
		jest.spyOn(uuid, "uuid").mockReturnValue("solo-field-id");

		const testSolo = renderer.create(<RichText {...nonRepeatableProps} />);
		expect(testSolo.toJSON()).toMatchSnapshot();
	});

	it("renders a repeating field if repeatable", () => {
		const repeatableProps = {
			modelSlug: "cats",
			field: {
				slug: "name",
				name: "Name",
				description: "The cat's real name.",
				value: ["<p>Kahn The Destroyer<p>"],
				isRepeatableRichText: true,
			},
		};

		// A dynamic ID would break snapshot testing, so we use a fixed one.
		jest.spyOn(uuid, "uuid").mockReturnValue("repeatable-field-id");

		const testRepeatable = renderer.create(
			<RichText {...repeatableProps} />
		);
		expect(testRepeatable.toJSON()).toMatchSnapshot();
	});
});
