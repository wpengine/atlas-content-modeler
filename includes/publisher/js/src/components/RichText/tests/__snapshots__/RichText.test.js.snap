// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RichText renders a repeating field if repeatable 1`] = `
Array [
  <label
    htmlFor="atlas-content-modeler[cats][name]"
  >
    Name
  </label>,
  <p
    className="help mb-4"
  >
    The cat's real name.
  </p>,
  <div
    className="css-15q5af2-RepeatingRichText"
    data-testid="rich-text-repeater-row"
  >
    <textarea
      defaultValue="<p>Kahn The Destroyer<p>"
      id="repeatable-field-id"
      name="atlas-content-modeler[cats][name][0]"
    />
  </div>,
  <div
    className="css-1tzuh7p-RepeatingRichText"
  >
    <button
      className="css-1c4y1ep-AddItemButton"
      data-testid="add-repeatable-row"
      onClick={[Function]}
      type="button"
    >
      <svg
        className="add add-small add-small-button"
        clipRule="evenodd"
        fillRule="evenodd"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          transform="translate(-20 -15)"
        >
          <path
            d="M33.333 33.333a.334.334 0 01-.333.334h-2a.334.334 0 01-.333-.334v-5h-5a.334.334 0 01-.334-.333v-2c0-.184.15-.333.334-.333h5v-5c0-.184.149-.334.333-.334h2c.184 0 .333.15.333.334v5h5c.184 0 .334.149.334.333v2c0 .184-.15.333-.334.333h-5v5z"
            fill="#7e5cef"
            fillRule="nonzero"
          />
        </g>
      </svg>
      <span>
        Add Item
      </span>
    </button>
  </div>,
]
`;

exports[`RichText renders a solo field if not repeatable 1`] = `
Array [
  <label
    htmlFor="atlas-content-modeler[cats][name]"
  >
    Name
  </label>,
  <p
    className="help mb-4"
  >
    The cat's real name.
  </p>,
  <textarea
    defaultValue="<p>Kahn The Destroyer</p>"
    id="solo-field-id"
    name="atlas-content-modeler[cats][name]"
  />,
]
`;
